rope/base/pyobjects.py

```diff
@@ -225,6 +228,9 @@ def get_module(self):
             current_object = current_object.parent
         return current_object
 
+    def get_name(self):
+        return None
```

Comment:
IIUC, this method and get_module() aren't supposed to be called, it's just a dummy implementation to satisfy mypy.

Maybe it should raise NotImplementedError or something like that instead to indicate that only the child class versions of these methods should actually get called.

Reply:
It's true that Rope probably never calls `PyDefinedObject.get_name`.

Imo, however, having to create these two methods indicates a more serious flaw in the Abstract classes.

At present, the draft PR #672 implements only step 1 of a multi-step project. This first step merely moves methods from the `Abstract` classes into what seem to me to be the more natural places, namely the `po.PyFunction` and `po.PyClass' classes (where `po` is my abbreviation for `rope.base.pyobjects`).

Step 2 of PR #672 proposes to make `PyDefinedObject` a subclass of `po.PyObject`. However, this is likely controversial and I have no great attachment to this change.

Oh what the heck. I may as well say here something that I was planning to save for a discussion.

After *weeks* of studying Rope's code I finally see the following:

1. Knowledge is a blessing as well as a curse :-)
   Knowledge helps experienced devs know what's important and what isn't.
2. For newbies the *lack* of knowledge is a curse.
    We can drown in details because we don't know which tidbit might be crucial.
    That's why your comments about inference were so important to me.
    I've just spent several hours revising [forked issue #3](https://github.com/edreamleo/ekr-fork-rope2/issues/3). I could not have done so without your help.
3. Imo, Rope uses a strange form of inheritance&mdash inheritance via mro.
   I understand what's happening now that I have spent days trying to revise it.
   But the class hierarchy will be difficult for newbies to understand without help.
   The Theory of Operation should explain in detail what is going on.
   The alternative is to use a more straightforward form of inheritance.
   The (closed) PR #664 demonstrates that such a simplification straightforward!
   The PR may have gone too far by eliminating the `Abstract` class entirely. It's an open question.

**Summary**

Newbies will struggle to understand the relations between Rope's most important classes.

PR #664 shows that the class hierarchy can be simplified. If this doesn't happen the Theory of Operation should contain a section about how Rope's classes are related.

Reply:
Meh. The phrase "inheritance by mro" is a fail.

The Aha is that the classes of interest, `po.PyObject`, `po.PyDefinedObject` and the three `Abstract` classes **are (mostly) used as mixin classes**. None of these classes are subclasses of the others, except that the `Abstract` classes are indeed subclasses of `po.PyObject`.

The existing class declarations are:
```python
class BuiltinClass(_BuiltinElement, pyobjects.AbstractClass):
class BuiltinFunction(_BuiltinElement, pyobjects.AbstractFunction):
class BuiltinModule(pyobjects.AbstractModule):
class Generator(pyobjects.AbstractClass):
class Iterator(pyobjects.AbstractClass):
class Lambda(pyobjects.AbstractFunction):
class AbstractClass(PyObject):
class AbstractFunction(PyObject):
class PyObject:
class PyDefinedObject:
class AbstractModule(PyObject):
class PyClass(PyDefinedObject, AbstractClass):
class PyFunction(PyDefinedObject, AbstractFunction):
class _PyModule(PyDefinedObject, AbstractModule):
```
In other words, `PyDefinedObject` *acts* like a `PyObject` even thought `PyDefinedObject` is not a subclass of `PyObject`. Why? Because all subclasses of `PyDefinedObject` are also subclasses of an `Abstract` class so that `PyObject` appears in the mro.

There is a weird kind of logic in this scheme. It's crazy, but imo it's not crazy enough to be right. PR #664 demonstrates that a simpler (more conventional) scheme is feasible.